_posts/2018-12-05-exploit-model.md

```diff
@@ -0,0 +1,184 @@
++-------------------+   +-------------------+
+<<<   7 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+For nabla, the attack surface to obtain code exec in the kernel would need to be from these 7 syscalls.
+
+To summarize, here's a side-by-side comparison of the models:
+
+```
+      GVISOR                   NABLA
+      ------                   -----
+
++--------+              
+|  p1a   |                                              APP
++--------+              +---------+
+<<< 300 syscalls  >>>   |   p2a   |          
++-------------------+   +-------------------+
+|      gvisor       |   |   nabla LibOS     |           LibOS
+|   (code exec)     |   |                   |            
++-------------------+   +-------------------+
+<<<  83 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+
+
+
+## But it's a memory safe langauge!
+
+So gvisor does two things, it reduces the attack surface to the kernel, and has an protection boundary which guards against the inputs that can be invoked with the 83 syscalls (the guarding is implicit in a way that the logic of the implementation would constrain the total number of syscalls invokable by a well behaved gvisor). Therefore, more work needs to be done by the attacker to bypass these guards.
+
+Some argue that gvisor is written in golang, and golang is memory safe. I believe that golang is a safer language from the PL standpoint, but it is hard to guarantee that the implementation is true to the language semantics. Often, bugs in implementation or performance trade-offs (this is rampant in the hardware work with [Spectre](https://meltdownattack.com/) variant bugs) are the result of this.
+
+As a bonus, here are two implementations of memory corruption POCs - one from an implementation bug, and one from a performance design standpoint.
+
+### Implementation bug
+
+The security research and CTF group that I am part of, [Plaid Parliament of Pwning (PPP)](https://pwning.net/), worked on showing code exec in the golang playground. You may view the blog post by Alex Reece [here](http://codearcana.com/posts/2013/04/23/exploiting-a-go-binary.html).
+
+### Design bug
+
+The bug here is due a data race of some types using multiword values (i.e. a type is represented by multiple words in byte) and a race in the garbage collector. 
+
+The vulnerability is detailed on STALKR's blog post ["Golang data races to break memory safety"](https://blog.stalkr.net/2015/04/golang-data-races-to-break-memory-safety.html) and discussion on the design and trade-offs are talked about in [Russ Cox's blogpost](https://research.swtch.com/gorace).
+
```

Comment:
Abrupt ending warning (I've been guilty of the same!) 
Need an ending summary line to the blogpost.